consensus/ibft/src/integration-test/java/tech/pegasys/pantheon/consensus/ibft/support/TestContextFactory.java

```diff
@@ -0,0 +1,264 @@
+import tech.pegasys.pantheon.ethereum.core.Wei;
+import tech.pegasys.pantheon.ethereum.db.WorldStateArchive;
+import tech.pegasys.pantheon.ethereum.mainnet.BlockHeaderValidator;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.util.bytes.BytesValue;
+import tech.pegasys.pantheon.util.uint.UInt256;
+
+import java.time.Clock;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Iterables;
+
+public class TestContextFactory {
+
+  private static class ControllerAndState {
+
+    private IbftController controller;
+    private IbftFinalState finalState;
+
+    public ControllerAndState(final IbftController controller, final IbftFinalState finalState) {
+      this.controller = controller;
+      this.finalState = finalState;
+    }
+
+    public IbftController getController() {
+      return controller;
+    }
+
+    public IbftFinalState getFinalState() {
+      return finalState;
+    }
+  }
+
+  public static final int EPOCH_LENGTH = 10_000;
+  public static final int BLOCK_TIMER_SEC = 3;
+  public static final int ROUND_TIMER_SEC = 12;
+
+  public static TestContext createTestEnvWithUtcClock(
+      final int validatorCount, final int indexOfFirstLocallyProposedBlock) {
+    return createTestEnvironment(
+        validatorCount, indexOfFirstLocallyProposedBlock, Clock.systemUTC());
+  }
+
+  public static TestContext createTestEnvironment(
+      final int validatorCount, final int indexOfFirstLocallyProposedBlock, final Clock clock) {
+
+    final NetworkLayout networkNodes =
+        NetworkLayout.createNetworkLayout(validatorCount, indexOfFirstLocallyProposedBlock);
+
+    final Block genesisBlock = createGenesisBlock(networkNodes.getValidatorAddresses());
+    final MutableBlockchain blockChain =
+        createInMemoryBlockchain(genesisBlock, IbftBlockHashing::calculateHashOfIbftBlockOnChain);
+
+    final KeyPair nodeKeys = networkNodes.getLocalNode().getNodeKeys();
+
+    // Use a stubbed version of the multicaster, to prevent creating PeerConnections etc.
+    final StubIbftMulticaster stubbedNetworkPeers = new StubIbftMulticaster();
+
+    final ControllerAndState controllerAndState =
+        createControllerAndFinalState(blockChain, stubbedNetworkPeers, nodeKeys, clock);
+
+    // Add each networkNode to the Multicaster (such that each can receive msgs from local node).
+    // NOTE: the remotePeers needs to be ordered based on Address (as this is used to determine
+    // the proposer order which must be managed in test).
+    final Map<Address, ValidatorPeer> remotePeers =
+        networkNodes
+            .getRemotePeers()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    NodeParams::getAddress,
+                    np ->
```

Comment:
This would be a lot more readable if the `np` abbreviation wasn't used. We can't see the type info with lambdas so it's not particularly clear what `np` is short for.

Reply:
done.